### PR TITLE
Support multiple sources for syntax tests.

### PR DESCRIPTION
**Pull request by ekpyron**
Implementation for syntax tests only, since the execution framework would need changes for supporting multiple sources in semantics tests.

### CODE REVIEW

**Comment by ekpyron** [comment]
The only import tests that can't yet be extracted are the remapping tests.
Remappings could be a syntax test setting, then they could be extracted as well... not sure if I add that to this PR as well, though...

File: test/libsolidity/syntaxTests/imports/shadowing_builtins_with_alias.sol [hunk 1]
```diff
@@ -0,0 +1,7 @@
+==== Source: B.sol ====
+contract C {}
+==== Source: b ====
+pragma solidity >=0.0;
+import {C as msg} from "B.sol";
+// ----
+// Warning: (B.sol:0-13): This declaration shadows a builtin symbol.
```
Comment (ekpyron) [inline test/libsolidity/syntaxTests/imports/shadowing_builtins_with_alias.sol:6]: The source locations here are in fact a bit off - I created #7273 to track that. I confirmed that it's not related to this PR, but also happens exactly like this with two "proper" source files.

**Comment by ekpyron** [comment]
Note that this *can't* deal with source names containing a ``:`` or starting with a digit - but to me it seems that these limitations are fine...
EDIT: See the parsing of source locations to ``SyntaxTestError``  with source names in ``SyntaxTest::parseExpectations``. If we decide that it'd be nice to support such source names, we could e.g. just require quotes around the source name, e.g. ``Warning ("42weird:source:name":1-4): ...``

File: test/TestCase.cpp [hunk 2]
```diff
@@ -80,7 +84,24 @@ pair<string, size_t> TestCase::parseSourceAndSettingsWithLineNumbers(istream& _s
...
 		else if (sourcePart)
-			source += line + "\n";
+		{
+			if (boost::algorithm::starts_with(line, sourceDelimiterStart) && boost::algorithm::ends_with(line, sourceDelimiterEnd))
+			{
+				if (!(currentSourceName.empty() && currentSource.empty()))
+					sources[currentSourceName] = std::move(currentSource);
+				currentSource = std::string();
+				size_t nameLength = line.size() - sourceDelimiterStart.size();
+				nameLength = nameLength >= sourceDelimiterEnd.size() ? nameLength - sourceDelimiterEnd.size() : 0;
```
Comment (chriseth) [inline test/TestCase.cpp]: Added this check to prevent underflow.

Comment (chriseth) [reply test/TestCase.cpp]: Ah wait a second, it's actually not necessary. Sorry...

File: test/libsolidity/SyntaxTest.cpp [hunk 3]
```diff
@@ -137,51 +146,65 @@ bool SyntaxTest::printExpectationAndError(ostream& _stream, string const& _lineP
...
-							sourceFormatting[i] = formatting::ORANGE_BACKGROUND_256;
-					}
-					else
-						sourceFormatting[i] = formatting::RED_BACKGROUND;
-			}
+	if (m_sources.empty())
+		return;
+
+	bool outputSourceNames = true;
+	if (auto it = m_sources.begin(); m_sources.size() == 1 && it->first == "")
```
Comment (chriseth) [inline test/libsolidity/SyntaxTest.cpp]:
```suggestion
	if (m_sources.size() == 1 && m_sources.begin()->first.empty())
```

**Review by chriseth — DISMISSED** [review]

File: test/libsolidity/SyntaxTest.cpp [hunk 4]
```diff
@@ -74,7 +75,10 @@ TestCase::TestResult SyntaxTest::run(ostream& _stream, string const& _linePrefix
 {
 	string const versionPragma = "pragma solidity >=0.0;\n";
 	compiler().reset();
-	compiler().setSources({{"", versionPragma + m_source}});
+	auto sourcesWithPragma = m_sources;
+	for(auto& source: sourcesWithPragma)
```
Comment (erak) [inline test/libsolidity/SyntaxTest.cpp]:
```suggestion
	for (auto& source: sourcesWithPragma)
```

File: test/libsolidity/SyntaxTest.cpp [hunk 5]
```diff
@@ -137,51 +146,65 @@ bool SyntaxTest::printExpectationAndError(ostream& _stream, string const& _lineP
...
+
+	bool outputSourceNames = true;
+	if (m_sources.size() == 1 && m_sources.begin()->first.empty())
+		outputSourceNames = false;
 
-		_stream << _linePrefix << sourceFormatting.front() << m_source.front();
-		for (size_t i = 1; i < m_source.length(); i++)
+	if (_formatted)
+	{
+		for(auto const& [name, source]: m_sources)
```
Comment (erak) [inline test/libsolidity/SyntaxTest.cpp]:
```suggestion
		for (auto const& [name, source]: m_sources)
```

File: test/libsolidity/SyntaxTest.cpp [hunk 6]
```diff
@@ -137,51 +146,65 @@ bool SyntaxTest::printExpectationAndError(ostream& _stream, string const& _lineP
...
+			vector<char const*> sourceFormatting(source.length(), formatting::RESET);
+			for (auto const& error: m_errorList)
+				if (error.sourceName == name && error.locationStart >= 0 && error.locationEnd >= 0)
+				{
+					assert(static_cast<size_t>(error.locationStart) <= source.length());
+					assert(static_cast<size_t>(error.locationEnd) <= source.length());
+					bool isWarning = error.type == "Warning";
+					for (int i = error.locationStart; i < error.locationEnd; i++)
+						if (isWarning)
+						{
```
Comment (erak) [inline test/libsolidity/SyntaxTest.cpp:172]: `{` and `}` are not necessary.

File: test/libsolidity/SyntaxTest.cpp [hunk 7]
```diff
@@ -137,51 +146,65 @@ bool SyntaxTest::printExpectationAndError(ostream& _stream, string const& _lineP
...
+
 	}
 	else
-	{
-		stringstream stream(m_source);
-		string line;
-		while (getline(stream, line))
-			_stream << _linePrefix << line << endl;
-	}
+		for(auto const& [name, source]: m_sources)
```
Comment (erak) [inline test/libsolidity/SyntaxTest.cpp]:
```suggestion
		for (auto const& [name, source]: m_sources)
```

File: test/TestCase.cpp [hunk 8]
```diff
@@ -80,7 +84,22 @@ pair<string, size_t> TestCase::parseSourceAndSettingsWithLineNumbers(istream& _s
 		else if (boost::algorithm::starts_with(line, settingsDelimiter))
 			sourcePart = false;
 		else if (sourcePart)
-			source += line + "\n";
+		{
+			if (boost::algorithm::starts_with(line, sourceDelimiterStart) && boost::algorithm::ends_with(line, sourceDelimiterEnd))
+			{
+				if (!(currentSourceName.empty() && currentSource.empty()))
+					sources[currentSourceName] = std::move(currentSource);
+				currentSource = std::string();
```
Comment (erak) [inline test/TestCase.cpp]:
```suggestion
				currentSource = string{};
```
or even

```suggestion
				currentSource = {};
```

File: test/TestCase.cpp [hunk 9]
```diff
@@ -95,12 +114,26 @@ pair<string, size_t> TestCase::parseSourceAndSettingsWithLineNumbers(istream& _s
 		else
 			throw runtime_error(string("Expected \"//\" or \"// ---\" to terminate settings and source."));
 	}
-	return make_pair(source, lineNumber);
+	sources[currentSourceName] = currentSource;
+	return make_pair(sources, lineNumber);
```
Comment (erak) [inline test/TestCase.cpp]:
```suggestion
	return {sources, lineNumber};
```

**Comment by chriseth** [comment]
Will fix!

File: test/libsolidity/SyntaxTest.cpp (hunk 6 above)
Comment (chriseth) [reply test/libsolidity/SyntaxTest.cpp:172]: I think they are, because of the else.

File: test/TestCase.cpp [hunk 10]
```diff
@@ -95,12 +114,26 @@ pair<string, size_t> TestCase::parseSourceAndSettingsWithLineNumbers(istream& _s
...
+	auto [sourceMap, lineOffset] = parseSourcesAndSettingsWithLineNumbers(_stream);
+	if (sourceMap.size() != 1)
+		BOOST_THROW_EXCEPTION(runtime_error("Expected single source definition, but got multiple sources."));
+	return {std::move(sourceMap.begin()->second), lineOffset};
 }
 
 string TestCase::parseSourceAndSettings(istream& _stream)
 {
-	return get<0>(parseSourceAndSettingsWithLineNumbers(_stream));
+	return parseSourceAndSettingsWithLineNumbers(_stream).first;
```
Comment (erak) [inline test/TestCase.cpp:136]: I would prefer to use `std::get` here. Using member access for `pair` is straightforward, but it occurred to me, that I had to turn a `pair` into a `tuple` with >2 fields, and using `std::get` helped to reduce refactoring effort. But it's also not a strong opinion ;)

File: test/libsolidity/SyntaxTest.cpp (hunk 6 above)
Comment (erak) [reply test/libsolidity/SyntaxTest.cpp:172]: Right. Overlooked the `else`.

**Review by erak — COMMENTED** [review]
Just a few tiny style-related comments. Otherwise it looks good!

**Review by chriseth — APPROVED** [review]

**Review by erak — APPROVED** [review]

**Pull request merged**